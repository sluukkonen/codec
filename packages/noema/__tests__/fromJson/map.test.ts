import { fromJson, integer, string } from "../../src/index.js"
import { expectParseFailure, expectParseSuccess } from "../helpers.js"

const stringToInt = fromJson.map(string, integer)

test("should parse valid arrays to a map", () => {
  expectParseSuccess(stringToInt, [], new Map())
  expectParseSuccess(stringToInt, [["a", 1]], new Map([["a", 1]]))
})

test("should not accept arrays containing invalid keys", () => {
  expectParseFailure(stringToInt, [[1, 1]])
})

test("should not accept arrays containing invalid values", () => {
  expectParseFailure(stringToInt, [["1", "1"]])
})

test("should not accept non-arrays", () => {
  expectParseFailure(stringToInt, null)
})

test("should not accept too short tuples", () => {
  expectParseFailure(stringToInt, [["a"]])
})

test("should not accept too long tuples", () => {
  expectParseFailure(stringToInt, [["a", 1, "extra"]])
})

test("should not accept arrays containing non-arrays", () => {
  expectParseFailure(stringToInt, [null])
})

test("should support complex key codecs", () => {
  const complex = fromJson.map(fromJson.bigint, string)
  expectParseSuccess(complex, [["1", "1"]], new Map([[1n, "1"]]))
})

test("should support complex value codecs", () => {
  const complex = fromJson.map(string, fromJson.bigint)
  expectParseSuccess(complex, [["1", "1"]], new Map([["1", 1n]]))
})
