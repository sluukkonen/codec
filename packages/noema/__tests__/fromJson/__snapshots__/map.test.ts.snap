// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not accept arrays containing invalid keys 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "string",
    "path": [
      0,
      0,
    ],
    "received": "number",
  },
]
`;

exports[`should not accept arrays containing invalid values 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "number",
    "path": [
      0,
      1,
    ],
    "received": "string",
  },
]
`;

exports[`should not accept arrays containing non-arrays 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "array",
    "path": [
      0,
    ],
    "received": "null",
  },
]
`;

exports[`should not accept non-arrays 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "array",
    "path": [],
    "received": "null",
  },
]
`;

exports[`should not accept too long tuples 1`] = `
[
  {
    "code": "too_long",
    "length": 3,
    "maxLength": 2,
    "path": [
      0,
    ],
  },
]
`;

exports[`should not accept too short tuples 1`] = `
[
  {
    "code": "too_short",
    "length": 1,
    "minLength": 2,
    "path": [
      0,
    ],
  },
]
`;
